lib/SILGen/SILGenPoly.cpp

```diff
@@ -392,12 +392,9 @@ ManagedValue Transform::transform(ManagedValue v,
     });
   }
 
-  // If the value is an optional, but the desired formal type isn't an
-  // optional, force it. We should only ever get here with values
-  // declared as implicitly unwrapped optionals because type checking
-  // should weed out cases involving plain optionals.
-  if (inputOTK != OTK_None && outputOTK == OTK_None) {
-    assert(inputOTK == OTK_ImplicitlyUnwrappedOptional);
```

Comment:
If the input is an optional and the output is Any, we would have hit the assertion. Or if the input is an optional of a hashable type and the output is AnyHashable. Probably the IUO forcing check should go at the very end, once the other conversions have been tried and failed.

Reply:
Yeah I was thinking the same. I will take a look at what's actually happening in the source compat test that failed.